### ISSUE TITLE
lldb can not print Option<i32>

### ISSUE THREAD

**Issue by jrmuizel** [rust-lang/rust#79530]
I tried this code:

```rust
{
    let x = Some(4);
    println!("Hello, world! {:?}", x);
}
```
Trying to print the value of `x` gives:
```
(core::option::Option<i32>) $0 = {}

```

This is with:
```
lldb-1200.0.41
Apple Swift version 5.3.1 (swiftlang-1200.0.41 clang-1200.0.32.8)
```
and
```
rustc 1.49.0-nightly (adef9da30 2020-10-13)
```

**Comment by tromey** [comment]
This is not a bug in rustc.  Instead, this case requires a patch to lldb.  The issue is that `Option` is a discriminated union, so the resulting DWARF looks like:
```
 <3><ad9>: Abbrev Number: 5 (DW_TAG_structure_type)
    <ada>   DW_AT_name        : (indirect string, offset: 0x876): Option<i32>
    <ade>   DW_AT_byte_size   : 8
    <adf>   DW_AT_alignment   : 4
 <4><ae0>: Abbrev Number: 7 (DW_TAG_variant_part)
    <ae1>   DW_AT_discr       : <0xae5>
...
```

However, the variant part construct isn't used by C or C++ (I don't know about Swift) and so stock lldb ignores it.

IIRC the rust lldb patches handled this ok.  So does gdb:
```
(gdb) n
4	    println!("Hello, world! {:?}", x);
(gdb) print x
$1 = core::option::Option<i32>::Some(4)
```

**Comment by jrmuizel** [comment]
@tromey, how hard would it be get an upstreamable fix for lldb out of your rust lldb patches?

**Comment by tromey** [comment]
> @tromey, how hard would it be get an upstreamable fix for lldb out of your rust lldb patches?

Sorry about the extreme delay on this response.  I forgot about it.  Anyway, it's hard to know the answer.  I feel that my track record getting things into upstream lldb is poor, and I don't know if that's due to something I did, or what.  Also, this is the kind of work I wouldn't want to do in my spare time.

**Comment by davxy** [comment]
@tromey do you know if there has been any progress on this topic? I think that the vscode-lldb guys may propose their patch to the upstream maybe... wdyt?

**Comment by VladimirMakaev** [comment]
Hi @tromey my employer is making contributions to upstream LLDB and we might be interested to get this fixed upstream as it'll improve our internal debugging experience. Would you be able to point me at the right place where this is fixed in a fork so I can engage with LLDB folks on this.  I'm myself not familiar with LLDB codebase so would appreciate any ideas you have on this.

**Comment by tromey** [comment]
Sorry again for the delay.  My changes were all in the rust lldb.  I'm sure they are still there in the history somewhere, though I don't know exactly where.  I'd imagine they would require some effort to update to the latest lldb.  Luckily the patches were not really that large.

**Comment by VladimirMakaev** [comment]
I've recently submitted a patch to LLDB to fix Rust enums which has got accepted https://reviews.llvm.org/D149213. If anybody has any feedback on this let me know. I'm planning to do a follow up contribution to synthetic providers  in rust-lang

**Comment by jgardona** [comment]
Tested here today with latest version, not working yet.
```
(lldb) p some
(core::option::Option<i32>)
```

**Comment by VladimirMakaev** [comment]
@jgardona What version of LLDB do you have? The fix should be in lldb 18.1+. You won't see this pretty printed though as the synthetic provider was never merged to rust-project. What you should see is all enum variants and their discriminants but that should let you understand what the value is (or allow to write a synth provider to do so)

**Comment by jgardona** [comment]
> @jgardona What version of LLDB do you have? The fix should be in lldb 18.1+. You won't see this pretty printed though as the synthetic provider was never merged to rust-project. What you should see is all enum variants and their discriminants but that should let you understand what the value is (or allow to write a synth provider to do so)

It worked. This is way better as we can see data inside and understand what it is.
Thank you :+1: 

![image](https://github.com/rust-lang/rust/assets/146892539/3d86abea-4f08-4225-ab19-5ac80f06f034)

**Comment by VladimirMakaev** [comment]
It’ll get nicer with the next release of rust compiler when you use ‘rust-lldb’ script. The new synthetic provider has just been merged in yesterday.
